### ISSUE TITLE
Extensible Identity Provider support

### ISSUE THREAD

**Issue by MJRichardson** [OctopusDeploy/Issues#2593]
Add support for using Azure Active Directory for authentication.

**Comment by slewis74** [comment]
Some of the functionality we're targeting as part of this implementation is:

- Extensibility.  The authentication providers will be loaded as extensions.
- Multiple providers.  We're aiming to add support for both Azure AD and GoogleApps out of the box.
- Multiple providers at once.  You'll be able to enable/disable each authentication provider independently.  E.g. you can enable Octopus Username/Password and Azure AD at the same time.
- "Auto Login".  In AD you'll be able to choose to force the NTML challenge, to save the user having to explicitly click the link.  Similarly, when using Azure AD or GoogleApps the user would be directed straight to the external authentication Url.

Octopus Username/Password and Active Directory will both still be supported, and we'll be addressing a couple of outstanding items related to Active Directory authentication.

**Comment by psyvision** [comment]
Looking forward to the "Auto Login" !!!

**Comment by johnkors** [comment]
How about targeting OIDC providers?

**Comment by slewis74** [comment]
Yep, the Azure AD and GoogleApps implementations are both OpenID Connect based, and they share lots of code.  

They'll be published as separate extensions for a couple of reasons but the main one is that we're supporting having multiple providers enabled at once.  The physical separation aligning with the conceptual separation then makes a few things easier.  There's also things like the custom UI assets for the Sign In page, where it makes sense for each extension to provide UI tailored to it's provider.

The intention is also to Open Source the extensions, so you could customise and replace our implementations if need be or you could write your own extension for another provider (reusing our base OpenID Connect code if that makes sense for the provider you're targeting).

**Comment by johnkors** [comment]
Yeah, Cool that.

I have my own OIDC provider (identityserver3 based), and would like to be able to use that for auth.

So, can I add a pic for it, provide my client_id & authority / .wellknown-openidconfig endpoint - and it's all set? 

Or will we have to write an extension for every instance of idsrv?

**Comment by slewis74** [comment]
We actually used idsvr for our initial POC and implementation, and the shared OpenID Connect code is based on that implementation, so you shouldn't have any issue with using idsvr.

The server and UI actually "ask the extensions" for the UI they want to contribute to the Sign In page, so you can absolutely provide custom pics and button styling etc.  For example, the 

```        protected override string LoginLinkHtml(string siteBaseUri)
        {
            return $"<div class='text-center'><a href='{siteBaseUri}/myextensions/authenticate'><img src=\"{siteBaseUri}/images/myextensions/btn_signin.png\"/></a></div>";
        }
```
We use Nancy, and the extensions will have the ability to contribute custom routes and static content as required.  There'll be blog posts and doco to go into this in more detail when it's released.

RE the .wellknown-configuration endpoint, the base OpenID Connect code has lots of plumbing to handle all of that for you.  To configure it all you require is the Issuer and ClientId (Scopes, ResponseType, ResponseMode etc have default values but can be overridden too).  The Issuer is the Url to your auth server and the plumbing reads the .wellknown-configuration and handles reading/caching the X509 certs for token validation and a few other things from there.

If I'm understanding your last question correctly, you shouldn't need an extension for every instance of idsrv.  The extensions have complete control of whatever config values they want to support, so for example if that extension wanted to support specifying a custom image for the login button it could do that.

**Comment by johnkors** [comment]
Perfect :) Looking forward to giving it a go!

**Comment by slewis74** [comment]
Release Note: an Azure AD authentication provider is included as a standard server extension

**Comment by DanielSundberg** [comment]
So this will be included in 3.5? Is there a full specification of what will be included in the 3.5 authentication work? When will 3.5 be released?
